### PR TITLE
1.14 Fluid rework.

### PR DESCRIPTION
**Pull request by gigaherz**
May or may not be what forge wants.

### CODE REVIEW

**Comment by KingLemming** [comment]
Well, I had been working on this offline but lemme see if I can't reconcile this with what I was thinking.

**Comment by mcenderdragon** [comment]
Looks good, does the fluid correctly render and flow ingame?

**Comment by gigaherz** [comment]
Yes, but with the caveat only the vanilla renderer is supported, no custom forge fluid model. I have been waiting for KL to see how things move forward.

File: src/main/java/net/minecraftforge/client/ForgeClientHandler.java [hunk 1]
```diff
@@ -24,27 +24,7 @@
 import net.minecraftforge.client.model.ModelLoader;
 import net.minecraftforge.common.ForgeMod;
 import net.minecraftforge.eventbus.api.SubscribeEvent;
-import net.minecraftforge.fluids.FluidRegistry;
 
 public class ForgeClientHandler
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/client/ForgeClientHandler.java]: This class is dead now, might as well kill it

File: patches/minecraft/net/minecraft/fluid/LavaFluid.java.patch [hunk 2]
```diff
@@ -0,0 +1,11 @@
+--- a/net/minecraft/fluid/LavaFluid.java
++++ b/net/minecraft/fluid/LavaFluid.java
+@@ -188,6 +188,8 @@
+       return 100.0F;
+    }
+ 
++   public net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid) { return net.minecraftforge.fluids.FluidAttributes.VANILLA_LAVA; }
```
Comment (ichttt) [inline patches/minecraft/net/minecraft/fluid/LavaFluid.java.patch]: Should stick an `@Override` before this, just to make sure

File: patches/minecraft/net/minecraft/fluid/WaterFluid.java.patch [hunk 3]
```diff
@@ -0,0 +1,11 @@
+--- a/net/minecraft/fluid/WaterFluid.java
++++ b/net/minecraft/fluid/WaterFluid.java
+@@ -94,6 +94,8 @@
+       return p_215665_5_ == Direction.DOWN && !p_215665_4_.func_207185_a(FluidTags.field_206959_a);
+    }
+ 
++   public net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid) { return net.minecraftforge.fluids.FluidAttributes.VANILLA_WATER; }
```
Comment (ichttt) [inline patches/minecraft/net/minecraft/fluid/WaterFluid.java.patch]: Same here, I'd use `@Override`

File: src/main/java/net/minecraftforge/fluids/IFluidTank.java [hunk 4]
```diff
@@ -1,33 +1,16 @@
-/*
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/fluids/IFluidTank.java]: Why was this removed?

File: src/main/java/net/minecraftforge/fluids/capability/templates/FluidTank.java [hunk 5]
```diff
@@ -0,0 +1,229 @@
+/*
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/fluids/capability/templates/FluidTank.java:1]: Forge License Header please?

File: src/main/java/net/minecraftforge/fluids/FluidAttributes.java [hunk 6]
```diff
@@ -54,17 +56,31 @@
...
+
     private static final Logger LOGGER = LogManager.getLogger();
 
     public static final int BUCKET_VOLUME = 1000;
 
+    public static final FluidAttributes EMPTY = new FluidAttributes("air", null, null, -1)
+            .setDensity(0).setTemperature(0).setLuminosity(0);
+    public static final FluidAttributes VANILLA_WATER = new FluidAttributes("water", LOCATION_WATER_STILL, LOCATION_WATER_FLOW, LOCATION_WATER_OVERLAY, -1);
+    public static final FluidAttributes VANILLA_LAVA = new FluidAttributes("lava", LOCATION_LAVA_STILL, LOCATION_LAVA_FLOW, -1)
+            /* TODO: set temperature and stuff */;
```
Comment (gigaherz) [inline src/main/java/net/minecraftforge/fluids/FluidAttributes.java]: Note to self: fix this before the PR gets merged.

**Review by ichttt — COMMENTED** [review]
Just some small things I found while reading, not a full review

File: src/main/java/net/minecraftforge/fluids/IFluidTank.java [hunk 7]
```diff
@@ -1,33 +1,16 @@
...
- * This library is distributed in the hope that it will be useful,
- * but WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
- * Lesser General Public License for more details.
- *
- * You should have received a copy of the GNU Lesser General Public
- * License along with this library; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
- */
-
```
Comment (gigaherz) [inline src/main/java/net/minecraftforge/fluids/IFluidTank.java:19]: This probably needs to be restored?

File: src/main/java/net/minecraftforge/fml/ModLoader.java [hunk 8]
```diff
@@ -212,7 +212,7 @@ private void accumulateErrors(List<ModLoadingException> errors) {
         if (containers.size() != modInfoMap.size()) {
             LOGGER.fatal(LOADING,"File {} constructed {} mods: {}, but had {} mods specified: {}",
                     modFile.getFilePath(),
-                    containers.size(), containers.stream().map(ModContainer::getModId).collect(Collectors.toList()),
+                    containers.size(), containers.stream().map(c -> c != null ? c.getModId() : "(null)").collect(Collectors.toList()),
```
Comment (gigaherz) [inline src/main/java/net/minecraftforge/fml/ModLoader.java:215]: @cpw I'm assuming this is not meant to contain nulls here. I made this change here to allow a completely opaque crash to become a meaninful crash, in case a Mod annotation does not have a matching entry in the mods.toml file.

**Review by gigaherz — COMMENTED** [review]
I can't "request changes" on my own PR, but these are things I think need to be looked at before we merge.

File: src/main/java/net/minecraftforge/fluids/IFluidTank.java (hunk 4 above)
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/IFluidTank.java]: I believe KL copypasted the entire file from his draft implementation elsewhere, and that draft didn't have the license headers in it.

File: patches/minecraft/net/minecraft/fluid/LavaFluid.java.patch (hunk 2 above)
Comment (gigaherz) [reply patches/minecraft/net/minecraft/fluid/LavaFluid.java.patch]: Good catch.

File: src/main/java/net/minecraftforge/fluids/FluidStack.java [hunk 9]
```diff
@@ -123,14 +124,14 @@ public final Fluid getFluid()
         return fluidDelegate.get();
     }
 
-    public String getLocalizedName()
+    public String getDisplayName()
```
Comment (pupnewfster) [inline src/main/java/net/minecraftforge/fluids/FluidStack.java]: Would it make more sense to have this be an `ITextComponent` like `Item#getDisplayName(ItemStack stack)` is?

File: patches/minecraft/net/minecraft/fluid/Fluid.java.patch [hunk 10]
```diff
@@ -18,5 +26,36 @@
...
++   }
++
++   /**
++    * Creates the fluid attributes object, which will contain all the extended values for the fluid that aren't part of the vanilla system.
++    * Do not call this from outside. To retrieve the values use {@link Fluid#getAttributes()}
++    */
++   protected net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid)
++   {
++      if (fluid instanceof EmptyFluid)
++         return net.minecraftforge.fluids.FluidAttributes.builder("water", null, null)
```
Comment (TeamSpen210) [inline patches/minecraft/net/minecraft/fluid/Fluid.java.patch]: Shouldn't this ID be `"empty"`, not `"water"`?
```suggestion
+         return net.minecraftforge.fluids.FluidAttributes.builder("empty", null, null)
```

Comment (gigaherz) [reply patches/minecraft/net/minecraft/fluid/Fluid.java.patch]: Yes. Nice catch.

File: src/main/java/net/minecraftforge/fluids/FluidStack.java (hunk 9 above)
Comment (gigaherz) [reply src/main/java/net/minecraftforge/fluids/FluidStack.java]: Done. Good idea.

**Comment by LexManos** [comment]
This PR, and it's sister [Recipe Rewrite](https://github.com/MinecraftForge/MinecraftForge/pull/6014) are both good to go IMO. And can be merged, I just want them merged at the same time/quickly together. As they are both major binary breaking changes.

Once these two are in, we'll do a sweep of basic bugs/prs and call it good for a RB.

File: src/main/java/net/minecraftforge/fluids/capability/templates/EmptyFluidHandler.java [hunk 11]
```diff
@@ -19,69 +19,47 @@
...
-    {
-        return null;
-    }
-
     @Override
-    public int getFluidAmount()
-    {
-        return 0;
-    }
+    public FluidStack getFluidInTank(int tank) { return null; }
```
Comment (MajorTuvok) [inline src/main/java/net/minecraftforge/fluids/capability/templates/EmptyFluidHandler.java:39]: What about a non-null policy similiar to ItemStack, with an isEmpty() chexk to find out, whether the FluidStack is empty?

**Comment by gigaherz** [comment]
Merged in b7e1cc5f6

**Comment by oOMitchOo** [comment]
Since issue #5153 is referenced here I would like to know, how this makes Forge added fluids placeable into 'waterloggable' blocks? It's my understanding that IWaterLoggable is hardcoded to work with Minecraft water, and since it translates to a boolean property (waterlogged) in the blocks, which can get waterlogged, there is not really a way to make your own fluids placeable into 'waterloggable' blocks right now?

**Comment by gigaherz** [comment]
That is correct. Waterlogged blocks are hardcoded to 1 fluid. In the case of vanilla blocks, they are hardcoded to water, so they won't accept your fluid. You could make new blocks that waterlog to other things, but in any case, you will only ever have a finite set of fluids, because the fluid system requires the property values be known ahead of time.
Unnoen made a mistake there. The issue should NOT have been closed.
